helpers/manifest-parser.js

```diff
@@ -0,0 +1,325 @@
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const ALLOWED_DISPLAY_VALUES = [
+  'fullscreen',
+  'standalone',
+  'minimal-ui',
+  'browser'
+];
+
+const ALLOWED_ORIENTATION_VALUES = [
+  'any',
+  'natural',
+  'landscape',
+  'portrait',
+  'portrait-primary',
+  'portrait-secondary',
+  'landscape-primary',
+  'landscape-secondary'
+];
+
+function parseString(raw, trim) {
+  let value;
+  let warning;
+
+  if (typeof raw === 'string') {
+    value = trim ? raw.trim() : raw;
+  } else {
+    if (raw !== undefined) {
+      warning = 'ERROR: expected a string.';
+    }
+    value = undefined;
+  }
+
+  return {
+    raw,
+    value,
+    warning
+  };
+}
+
+function parseURL(raw) {
+  // TODO: resolve url using baseURL
+  // var baseURL = args.baseURL;
+  // new URL(parseString(raw).value, baseURL);
+  return parseString(raw, true);
+}
+
+function parseColor(raw) {
+  let color = parseString(raw);
+
+  if (color.value === undefined) {
+    return color;
+  }
+
+  // TODO(bckenny): validate color, but for reals
+  // possibly pull in devtools/front_end/common/Color.js
+  // If style.color changes when set to the given color, it is valid. Testing
+  // against 'white' and 'black' in case of the given color is one of them.
+  // var dummy = document.createElement('div');
+  // dummy.style.color = 'white';
+  // dummy.style.color = color.value;
+  // if (dummy.style.color !== 'white') {
+  //   return color;
+  // }
+  // dummy.style.color = 'black';
+  // dummy.style.color = color.value;
+  // if (dummy.style.color !== 'black') {
+  //   return color;
+  // }
+  return color;
+
+  // color.value = undefined;
+  // color.warning = 'ERROR: color parsing failed';
+
+  // return color;
+}
+
+function parseName(jsonInput) {
+  return parseString(jsonInput.name, true);
+}
+
+function parseShortName(jsonInput) {
+  return parseString(jsonInput.short_name, true);
+}
+
+function parseStartUrl(jsonInput) {
+  // TODO: parse url using manifest_url as a base (missing).
+  // start_url must be same-origin as Document of the top-level browsing context.
+  return parseURL(jsonInput.start_url);
+}
+
+function parseDisplay(jsonInput) {
+  let display = parseString(jsonInput.display, true);
+
+  if (display.value && ALLOWED_DISPLAY_VALUES.indexOf(display.value.toLowerCase()) === -1) {
+    display.value = undefined;
```

Comment:
It's from the web app manifest spec. It's what happens most of the time when you have an unrecognized value (though not always). Dropping the value at a higher level is an interesting idea, but it's going to be more state to track